### ISSUE TITLE
[PG] PreparedQuery results in `ambiguous_parameter` error (42P08)

### ISSUE THREAD

**Issue by rgmz** [eclipse-vertx/vertx-sql-client#1103]
### Version

4.2.1

### Context

I encountered this error while refactoring a query to avoid unnecessary updates ([similar to this StackOverflow post](https://dba.stackexchange.com/a/282199)):
```sql
-- Before (works)
UPDATE repository
SET name = $2, description = $3
WHERE id = $1;

-- After (doesn't work)
UPDATE repository
SET name = $2, description = $3
WHERE id = $1 AND (repository.*) IS DISTINCT FROM ($1, $2, $3);
```

Running this query with `PgPool.preparedQuery(...).execute(Tuple.of(...))` results in the following error:
```
2021-12-09 10:39:14,228 ERROR [com.example] (main @coroutine#2) CON: io.vertx.pgclient.PgException: ERROR: could not determine data type of parameter $2 (42P08)
	at io.vertx.pgclient.impl.codec.ErrorResponse.toException(ErrorResponse.java:31)
	at io.vertx.pgclient.impl.codec.PrepareStatementCommandCodec.handleErrorResponse(PrepareStatementCommandCodec.java:90)
	at io.vertx.pgclient.impl.codec.PgDecoder.decodeError(PgDecoder.java:246)
	at io.vertx.pgclient.impl.codec.PgDecoder.decodeMessage(PgDecoder.java:132)
	at io.vertx.pgclient.impl.codec.PgDecoder.channelRead(PgDecoder.java:112)
```

This is presumably because the client [creates a prepared statement without explicit data types](https://www.postgresql.org/docs/current/sql-prepare.html) and Postgres tries to infer the type "...from the context in which the parameter is first referenced", which I believe is actually in `WHERE` (ambiguous) and not in `SET`.

### Expected

Ideally it would be possible to either create a prepared statement with explicit data types from a tuple ([`PrepareStatementCommand`](https://github.com/eclipse-vertx/vertx-sql-client/blob/e4bf6161be606a313c6a7dce5add06c11aed70fd/vertx-sql-client/src/main/java/io/vertx/sqlclient/impl/command/PrepareStatementCommand.java) and [`PrepareStatementCommandCodec`](https://github.com/eclipse-vertx/vertx-sql-client/blob/9ebe55e0d9ed1b5a27378279d470ea750b7d70b7/vertx-pg-client/src/main/java/io/vertx/pgclient/impl/codec/PrepareStatementCommandCodec.java#L46-L48)) so the statement is created as:

```diff
-PREPARE foo AS
+PREPARE foo (bigint, text, text) AS
UPDATE repository
SET name = $2, description = $3
WHERE id = $1 AND (repository.*) IS DISTINCT FROM ($1, $2, $3);
```

Perhaps `42P08` should be added to the list of [indeterminate error codes](https://github.com/eclipse-vertx/vertx-sql-client/blob/d0d4c557c88a13dbc508033e3a36ce200213843b/vertx-pg-client/src/main/java/io/vertx/pgclient/impl/PgSocketConnection.java#L167-L173) as well.

### Extra

As discussed in previous issues (#699 and #851), not defining explicit data types is an explicit choice:

> We do not provide the parameters because some queries might use a different parameter types at execution time and that would result in a cast error in PG (e.g using a function that take any parameter as input, if you bind it with integer, then it will fail for a string).

**Comment by vietj** [comment]
can you show your code ?

**Comment by vietj** [comment]
if you are using `client.preparedQuery(the_query).execute(arg1, arg2, ...)` when this error is encountered then the client will use the types defined by arg1, ... and use them to re-prepare the query.

**Comment by vietj** [comment]
ah right I didn't see the remark about prepared codes that is necessary for what I said, so indeed I think that should be done

**Comment by vietj** [comment]
it seems also one solution is to force the type in the query, e.g

```
UPDATE repository
SET name = $2::varchar, description = $3::varchar
WHERE id = $1::varchar AND (repository.*) IS DISTINCT FROM ($1::varchar, $2::varchat, $3::varchar);
```

**Comment by vietj** [comment]
@rgmz 4.2.2 will be released this week as soon as Netty 4.1.72.Final is released
